source_code/main_mcu/src/LOGIC/logic_user.c

```diff
@@ -1837,154 +1837,270 @@ RET_TYPE logic_user_ask_for_credentials_keyb_output(uint16_t parent_address, uin
+                            state_machine = 0;
+                        }
+                        else
+                        {
+                            /* No login, no multiple interfaces connected, leave function */
+                            return RETURN_BACK;
+                        }
                     }
-                    else
-                    {
-                        /* No login, no multiple interfaces connected, leave function */
-                        return RETURN_BACK;                        
-                    }                 
                 }
                 else
                 {
-                    state_machine--;
+                    state_machine--; //State machine is now at password
                 }
             }
             else
             {
                 if (prompt_return == MINI_INPUT_RET_YES)
                 {
                     aux_mcu_message_t* typing_message_to_be_sent;
-                    BOOL prev_gen_credential_flag = FALSE;
-                    
-                    /* Check for previous generation password */
-                    if ((temp_cnode.flags & NODEMGMT_PREVGEN_BIT_BITMASK) != 0)
-                    {
-                        prev_gen_credential_flag = TRUE;
-                    }
-                    
-                    /* Decrypt password. The field just after it is 0 */
-                    logic_encryption_ctr_decrypt((uint8_t*)temp_cnode.password, temp_cnode.ctr, MEMBER_SIZE(child_cred_node_t, password), prev_gen_credential_flag);
-                    
-                    /* If old generation password, convert it to unicode */
-                    if (prev_gen_credential_flag != FALSE)
-                    {
-                        _Static_assert(MEMBER_SIZE(child_cred_node_t, password) >= NODEMGMT_OLD_GEN_ASCII_PWD_LENGTH*2 + 2, "Backward compatibility problem");
-                        utils_ascii_to_unicode((uint8_t*)temp_cnode.password, NODEMGMT_OLD_GEN_ASCII_PWD_LENGTH);
-                        temp_cnode.cust_char_password[NODEMGMT_OLD_GEN_ASCII_PWD_LENGTH] = 0;
-                    }
-                        
-                    /* Type shortcut if specified */
-                    if ((shortcut_sent == FALSE) && ((keys_to_send_before_login & (LF_ENT_KEY_MASK|LF_CTRL_ALT_DEL_MASK)) != 0))
+
+                    /* Decrypt TOTPsecret */
+                    logic_encryption_ctr_decrypt(temp_cnode.TOTP.TOTPsecret_ct, temp_cnode.TOTP.TOTPsecret_ctr, MEMBER_SIZE(TOTP_cred_node_t, TOTPsecret), FALSE);
+
+                    /* Generate TOTP */
+                    cust_char_t TOTP_str[LOGIC_GUI_TOTP_STR_LEN];
+                    memset(TOTP_str, 0, sizeof TOTP_str);
+                    logic_encryption_generate_totp(temp_cnode.TOTP.TOTPsecret_ct, temp_cnode.TOTP.TOTPsecretLen, temp_cnode.TOTP.TOTPnumDigits, temp_cnode.TOTP.TOTPtimeStep, TOTP_str, LOGIC_GUI_TOTP_STR_LEN);
+                    uint16_t TOTP_len = utils_strlen(TOTP_str);
```

Comment:
I checked logic_encryption_generate_totp for issues but for good code practises, could change from strlen to utils_strnlen ?

Reply:
:+1: